### PR TITLE
repr(transparent): it's fine if the one non-1-ZST field is a ZST

### PR DESCRIPTION
**Pull request by RalfJung**
This code currently gets rejected:
```rust
#[repr(transparent)]
struct MyType([u16; 0])
```
That clearly seems like a bug to me: `repr(transparent)` [got defined ](https://github.com/rust-lang/rust/issues/77841#issuecomment-716575747) as having any number of 1-ZST fields plus optionally one more field; `MyType` clearly satisfies that definition.

This PR changes the `repr(transparent)` logic to actually match that definition.

### CODE REVIEW

**Comment by RalfJung** [comment]
Nominating for T-lang since this is a change in public behavior on stable -- but we already [had an FCP](https://github.com/rust-lang/rust/issues/77841#issuecomment-716575747) blessing this so I don't think we need a full FCP again.

**Comment by fmease** [comment]
That's great, when working on a fix for #90435 (which is now sadly blocked on #114952 I guess), I've noticed that

```rs
#[repr(transparent)] struct S(_1Zst, Zst);
struct _1Zst;
#[repr(align(2))] struct Zst;
```

currently doesn't compile. This should be fixed by this PR, right?

**Comment by RalfJung** [comment]
Yes that should be fixed by this PR.

**Comment by scottmcm** [comment]
I agree that this doesn't need another lang sign-off since it's what was explicitly FCP'd before.  (Sadly we didn't get to it in the meeting today, so I'm only speaking for me.)

**Comment by scottmcm** [comment]
Since this is opening up something new, could we fix https://github.com/rust-lang/rust/issues/90435 & https://github.com/rust-lang/rust/issues/90435 *for this case* as part of this change?

(Fixing those issues for existing code is harder, but if they didn't compile before maybe we can do it here?)

**Comment by fmease** [comment]
I've opened #115439 to fix #90435.

**Comment by RalfJung** [comment]
> Since this is opening up something new, could we fix https://github.com/rust-lang/rust/issues/90435 & https://github.com/rust-lang/rust/issues/90435 for this case as part of this change?

I'm not sure if this has to be entangled with rustdoc changes. That's an entirely orthogonal discussion IMO.

**Comment by scottmcm** [comment]
We discussed this briefly in the lang triage meeting today, and the meeting was in agreement that the previous FCP applies, and that we're happy to see this addressed.

**Comment by RalfJung** [comment]
@TaKO8Ki the ball is in your court then. :)

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
LGTM

@bors r+

**Pull request merged**
